### PR TITLE
Torchao

### PR DESCRIPTION
**Pull request by drisspg**
# Summary
Added TorchAO quantization support to vLLM. 

The required dependencies have been merged:
- [x] [[AO PR](https://github.com/pytorch/ao/pull/1875)](https://github.com/pytorch/ao/pull/1875)
- [x] [[HF PR](https://github.com/huggingface/transformers/pull/36526)](https://github.com/huggingface/transformers/pull/36526)

## Some discussion points
- I added a mechanism for doing on the fly quant via overriding the hf quant config. I am okay with removing this if we don't think this is a valuable workflow
- This requires latest TorchAO ( pip wheel not released yet, and same for transformers) 
- I have tested w/ different settings of TP and they all appear to work so far. I have also tested w/ gemma, llama, opt, and qwq -> all good
- There is a known issue w/ subclasses and how VLLM integrates w/ compile. I worked on this for a while w/ Brian on sunday seeing if we could patch it quickly. We will need to make sure that vllm uses the aotatuograd wrapping in its cache to ensure that it correctly unwraps subclasses. But for now, compiling once and removing the cache or disabliling cache works -> albeit slow start up every time so I do think this is important to fix.
- I haven't look to indepth to v1 engine so not sure if things need to be changed when that becomes default

### Other attempt
I also tried a different way of implementing this: https://github.com/vllm-project/vllm/pull/14393. Thesis, is that if the workflow works w/ regular tensors than it should work with subclasses. This fail quickly in two places -> The QKV module re-write needs to be able to fill a regular tensor w/ sub aqt tensors. I replaced this copy_ slices with cat to create one big subclass tensor first. Second, calling copy_on params.data is kinda of frowned upon - specifically breaks subclasses. I replaced these calls w/ swap_tensor. I had to do the qkv cat rewrite in model def land and I feel if we have to do this for every model def that is not a good outcome.

Perhaps its possible to roll the swap tensor changes into the weight loader.

### CODE REVIEW

File: tests/quantization/test_torchao.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+                assert isinstance(
+                    component,
+                    AffineQuantizedTensor), f"{name} is not quantized"
+
+
+@pytest.mark.parametrize("model", MODELS)
+@pytest.mark.parametrize("dtype", DTYPE)
+@pytest.mark.parametrize("config", CONFIGS)
+def test_torchao(vllm_runner, model, dtype, config):
+    hf_override_dict = {
```
Comment (drisspg) [inline tests/quantization/test_torchao.py:50]: @mgoin Still some work to be done on this PR but here is the hf_overrdide path? 

I do think its kind of nice to be able to quantize on the fly but maybe that isn't as important for most users? If so could be worth revisiting maybe after this PR for better UX

File: vllm/model_executor/layers/quantization/torchao.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        set_weight_attrs(weight, extra_weight_attrs)
+
+    def process_weights_after_loading(self, layer: Module) -> None:
+        # We aren't loading a quantized weight, we quantize on the fly
+        if self.quant_config.runtime_quant:
+            torchao_config = self.quant_config.torchao_config
+            layer.weight = torchao_quantize_param_data(layer.weight,
+                                                       torchao_config)
+
+    # @torch.compile(mode="max-autotune-no-cudagraphs")
```
Comment (drisspg) [inline vllm/model_executor/layers/quantization/torchao.py:128]: Still need to figure out whats the best way to plum compile settings through

File: tests/quantization/test_torchao.py (hunk 1 above)
Comment (mgoin) [reply tests/quantization/test_torchao.py:50]: I see, thanks for writing these examples. 

This does seem clunky, and I would like it to be easy for users to specify runtime quants if wanted. 

When I proposed the hf_override path, my thought was we don't need to keep the usage so literal to what would need to be in a proper config.json. I think it would be okay if we used the arg to sneak in some special tags/shortcuts, like the torchao_config had in the original PR. If we can interpret those tags in the right place, we can expand them into proper quant_configs. I'll try thinking a bit more tomorrow, let me know if you have ideas

File: vllm/model_executor/layers/quantization/torchao.py (hunk 2 above)
Comment (ProExpertProg) [reply vllm/model_executor/layers/quantization/torchao.py:128]: Which `torch.compile` settings need to be passed through here?

Comment (drisspg) [reply vllm/model_executor/layers/quantization/torchao.py:128]: We have this helper which sets some global state: https://github.com/vllm-project/vllm/pull/14231/files#diff-797bf189c4ed30c72616d9da827c9c4534784128fe4d1a7e0ddd435d25009267R20

I want to ensure that this isn't being imported and ran by default 

Sidenote:
Is this pattern adding a torch.compile on the apply func correct?

Comment (ProExpertProg) [reply vllm/model_executor/layers/quantization/torchao.py:128]: Most likely no. We have a model-wide torch compile wrapper so unless this is used outside the forward method or in models that don't support torch.compile, this is probably not right.

Could you elaborate on the use case here a bit more?

Comment (drisspg) [reply vllm/model_executor/layers/quantization/torchao.py:128]: Hey sorry I ended up adding specific tests for compile to ensure that it works accordingly and removed the intial line.

Most things appear to be working except that the inductor cache currently doens't have the correct AOTautograd calling wrapper to correctly unpack tensor subclasses. Spoke with Richard about this

File: vllm/model_executor/layers/quantization/torchao_utils.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+                                torchao_config: Any) -> torch.nn.Parameter:
+    """Quantize a Tensor with torchao quantization specified by torchao_config
+
+    Args:
+       `param`: weight parameter of the linear module
+       `torchao_config`: type of quantization and their arguments we want to
+        use to quantize the Tensor
+    """
+    try:
+        from torchao.core.config import AOBaseConfig
```
Comment (vkuzo) [inline vllm/model_executor/layers/quantization/torchao_utils.py:21]: IMO we should move `AOBaseConfig` to `torchao/__init__.py` and import from there

File: docs/source/features/quantization/torchao.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+pip install --pre torchao --index-url https://download.pytorch.org/whl/nightly/cu124  # or other cuda versions like cu121
+```
+
+You can quantize your own huggingface model with torchao, e.g. [transformers](https://huggingface.co/docs/transformers/main/en/quantization/torchao) and [diffusers](https://huggingface.co/docs/diffusers/en/quantization/torchao), and save the checkpoint to huggingface hub like [this](https://huggingface.co/jerryzh168/llama3-8b-int8wo) either with the following example code:
+
+```python
+import torch
+from transformers import TorchAoConfig, AutoModelForCausalLM, AutoTokenizer
+
+model_name = "meta-llama/Meta-Llama-3-8B"
```
Comment (drisspg) [inline docs/source/features/quantization/torchao.md:19]: Need to update this markdown

File: vllm/engine/arg_utils.py [hunk 5]
```diff
@@ -136,7 +136,9 @@ class EngineArgs:
     rope_theta: Optional[float] = None
     hf_overrides: Optional[HfOverrides] = None
     tokenizer_revision: Optional[str] = None
+
```
Comment (houseroad) [inline vllm/engine/arg_utils.py:139]: Remove empty lines here?

File: docs/source/features/quantization/torchao.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+tokenizer.push_to_hub(hub_repo)
+quantized_model.push_to_hub(hub_repo, safe_serialization=False)
+```
+
+or with [spaces](https://huggingface.co/spaces/medmekk/TorchAO_Quantization).
+
+We can then test latency for the quantized model with vllm:
+
+```console
+python benchmarks/benchmark_latency.py --input-len 256 --output-len 256 --model jerryzh168/llama3-8b-int8wo --batch-size 1 --quantization torchao --torchao-config int8wo
```
Comment (houseroad) [inline docs/source/features/quantization/torchao.md:36]: wondering what's the baseline?

Comment (houseroad) [reply docs/source/features/quantization/torchao.md:36]: Also wondering ht eaccuracy part.

Comment (jerryzh168) [reply docs/source/features/quantization/torchao.md:36]: @houseroad Accuracy impact is different for each model/task, but from experience generally 8 bit (int8/float8) and even 4 bit is fine most of the time: https://neuralmagic.com/blog/we-ran-over-half-a-million-evaluations-on-quantized-llms-heres-what-we-found/. we also have accuaracy preserving techniques like hqq, gptq to recover accuaracy in PTQ, and also have QAT as well. So it's a bit separate from our integration here.

File: tests/quantization/test_torchao.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        "quant_method": "torchao",
+        "quant_type": {
+            "default": config_to_dict(config)
+        },
+        "runtime_quant": True
```
Comment (jerryzh168) [inline tests/quantization/test_torchao.py:57]: this can be removed I think

**Comment by drisspg** [comment]
cc @mgoin @tlrmchlsmth Removed the runtime quant option and simplified the readme, could I get another review

File: docs/source/features/quantization/torchao.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
+(torchao)=
+
+# TorchAO
+
+TorchAO is an architecture optimization library for PyTorch, it provides high performance dtypes, optimization techniques and kernels for inference and training, featuring composability with native PyTorch features like torch.compile, FSDP etc.. Some benchmark numbers can be found [here](https://github.com/pytorch/ao/tree/main/torchao/quantization#benchmarks).
+
+We recommend installing the latest torchao nightly with
+
+```console
+pip install --pre torchao>=10.0.0 --index-url https://download.pytorch.org/whl/nightly/cu126  # or other cuda versions like cu128
```
Comment (mgoin) [inline docs/source/features/quantization/torchao.md:10]: You are missing a closing ``` here

File: vllm/model_executor/layers/quantization/torchao.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
+# SPDX-License-Identifier: Apache-2.0
+import importlib.metadata
+from typing import Any, Dict, List, Optional
+
+import torch
+import torch.nn.functional as F
+from packaging import version
+from torch.nn.parameter import Parameter
+from torchao.core.config import AOBaseConfig
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/torchao.py:9]: Can you move this into the config class so it can be lazy-imported?

**Review by mgoin — COMMENTED** [review]
Very clean implementation! I think this is good to go with a few comments

**Review by tlrmchlsmth — APPROVED** [review]
This looks good to me as well.

I did have one question, mostly out of curiosity: I heard that there was an issue with the way vLLM uses torch.compile, which disallows subclassing `torch.Tensor`, which got in the way of the torchao integration. How was that resolved?

**Comment by drisspg** [comment]
@tlrmchlsmth This is a really good point to bring up!  I touched on this briefly in the summary but worth reiterating.

State of the world today:
If you compile w/ `compilation_config=CompilationLevel.PIECEWISE` on the first go around everything works as expected w/ subclasses. On second invocation though, the  vllm compile cache will see a cache hit and attempt to re-run with the saved inductor artifacts. However these artifacts do not save the AOTDispatcher runtime wrapper metadata which is required for tensor-subclasses today. 

The current plan on record is to that when we enable inductor.compile to be its own public API this will include the AOTDispatcher and should fix this problem.

The work around today is `os.environ["VLLM_DISABLE_COMPILE_CACHE"] = "1"` which feels bad.. wasn't sure if we should very publically document this.

cc @zou3519 if I butchered any of this

File: vllm/model_executor/layers/quantization/torchao.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+from vllm.model_executor.layers.quantization.base_config import (
+    QuantizationConfig)
+from vllm.model_executor.layers.quantization.torchao_utils import (
+    torchao_quantize_param_data)
+from vllm.model_executor.utils import set_weight_attrs
+
+
+class TorchAOConfig(QuantizationConfig):
+    """Config class for torchao."""
+    from torchao.core.config import AOBaseConfig
```
Comment (ProExpertProg) [inline vllm/model_executor/layers/quantization/torchao.py:20]: Won't this throw an `ImportError` if torchao is not installed? `TorchAOConfig` is imported from `__init__` so this will always run unless I'm missing something

Comment (drisspg) [reply vllm/model_executor/layers/quantization/torchao.py:20]: you are right, will fix

**Review by ProExpertProg — APPROVED** [review]

File: tests/quantization/test_torchao.py [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+
+def test_pre_quantized_model(vllm_runner):
+    with vllm_runner("drisspg/float8_dynamic_act_float8_weight-opt-125m",
+                     quantization="torchao",
+                     dtype="bfloat16",
+                     compilation_config=CompilationLevel.PIECEWISE) as llm:
+        output = llm.generate_greedy(["The capital of France is"],
+                                     max_tokens=32)
+    assert output
+    print(output)
```
Comment (houseroad) [inline tests/quantization/test_torchao.py:24]: maybe length of output > 0 and maybe check Paris is part of it?

**Comment by drisspg** [comment]
Hey @tlrmchlsmth is there anything else needed?

**Review by tlrmchlsmth — APPROVED** [review]
PR looks ready to go, thanks for the ping @drisspg 🙏
